### PR TITLE
Add "param" substitution in tags

### PR DESCRIPTION
**Pull request by tahsinkose**
This PR implements the functionality proposed in #723. I only tested with `double`, `int` and `string` types. Actually, it converts any specified parameter of YAML into string to conform to the native behavior in `roslaunch`. It might require additional testing for different use cases.

### CODE REVIEW

**Comment by dirk-thomas** [comment]
The current patch contains numerous unrelated changes. Please remove those and keep the diff to a minimum. Please also consider to add a test for the new feature.

**Comment by tahsinkose** [comment]
> The current patch contains numerous unrelated changes. Please remove those and keep the diff to a minimum. Please also consider to add a test for the new feature.

I will be focusing on this ASAP.

**Comment by dirk-thomas** [comment]
> I will be focusing on this ASAP.

@tahsinkose Friendly ping.

**Comment by tahsinkose** [comment]
> The current patch contains numerous unrelated changes. Please remove those and keep the diff to a minimum. Please also consider to add a test for the new feature.

Could you please point out where the unrelated changes occur?

File: ros_comm/package.xml [hunk 1]
```diff
@@ -1,6 +1,6 @@
 <package>
   <name>ros_comm</name>
-  <version>1.14.3</version>
+  <version>1.14.4</version>
```
Comment (dirk-thomas) [inline ros_comm/package.xml]: Unrelated change. Should be be reverted.

File: tools/roslaunch/src/roslaunch/loader.py [hunk 2]
```diff
@@ -413,6 +413,7 @@ def load_rosparam(self, context, ros_config, cmd, param, file_, text, verbose=Tr
                 # #3162: if there is no YAML, load() will return an
                 # empty string.  We want an empty dictionary instead
                 # for our representation of empty.
+
```
Comment (dirk-thomas) [inline tools/roslaunch/src/roslaunch/loader.py]: Unrelated change. Should be be reverted.

File: tools/roslaunch/src/roslaunch/loader.py [hunk 3]
```diff
@@ -429,13 +430,12 @@ def load_rosparam(self, context, ros_config, cmd, param, file_, text, verbose=Tr
             # 'param' attribute is required for non-dictionary types
             if not param and type(data) != dict:
                 raise ValueError("'param' attribute must be set for non-dictionary values")
-
```
Comment (dirk-thomas) [inline tools/roslaunch/src/roslaunch/loader.py]: Unrelated change. Should be be reverted.

File: tools/roslaunch/src/roslaunch/loader.py [hunk 4]
```diff
@@ -429,13 +430,12 @@ def load_rosparam(self, context, ros_config, cmd, param, file_, text, verbose=Tr
...
             self.add_param(ros_config, full_param, data, verbose=verbose)
 
         else:
             raise ValueError("unknown command %s"%cmd)
 
-
+
```
Comment (dirk-thomas) [inline tools/roslaunch/src/roslaunch/loader.py]: Unrelated change. Should be be reverted.

File: tools/roslaunch/src/roslaunch/substitution_args.py [hunk 5]
```diff
@@ -369,6 +386,7 @@ def resolve_args(arg_str, context=None, resolve_anon=True, filename=None):
     @rtype:  str
     @raise SubstitutionException: if there is an error resolving substitution args
     """
+
```
Comment (dirk-thomas) [inline tools/roslaunch/src/roslaunch/substitution_args.py]: Unrelated change. Should be be reverted.

File: tools/roslaunch/src/roslaunch/substitution_args.py [hunk 6]
```diff
@@ -393,16 +412,19 @@ def resolve_args(arg_str, context=None, resolve_anon=True, filename=None):
     return resolved
 
 def _resolve_args(arg_str, context, resolve_anon, commands):
-    valid = ['find', 'env', 'optenv', 'dirname', 'anon', 'arg']
+    valid = ['find', 'env', 'optenv', 'dirname', 'anon', 'arg','param']
     resolved = arg_str
     for a in _collect_args(arg_str):
+
```
Comment (dirk-thomas) [inline tools/roslaunch/src/roslaunch/substitution_args.py]: Unrelated change. Should be be reverted.

File: tools/roslaunch/src/roslaunch/substitution_args.py [hunk 7]
```diff
@@ -393,16 +412,19 @@ def resolve_args(arg_str, context=None, resolve_anon=True, filename=None):
     return resolved
 
 def _resolve_args(arg_str, context, resolve_anon, commands):
-    valid = ['find', 'env', 'optenv', 'dirname', 'anon', 'arg']
+    valid = ['find', 'env', 'optenv', 'dirname', 'anon', 'arg','param']
     resolved = arg_str
     for a in _collect_args(arg_str):
+        
         splits = [s for s in a.split(' ') if s]
+        #print("splits: %s"%(splits))
```
Comment (dirk-thomas) [inline tools/roslaunch/src/roslaunch/substitution_args.py]: Unrelated change. Should be be reverted.

File: tools/roslaunch/src/roslaunch/substitution_args.py [hunk 8]
```diff
@@ -393,16 +412,19 @@ def resolve_args(arg_str, context=None, resolve_anon=True, filename=None):
...
         if not splits[0] in valid:
             raise SubstitutionException("Unknown substitution command [%s]. Valid commands are %s"%(a, valid))
         command = splits[0]
         args = splits[1:]
         if command in commands:
             resolved = commands[command](resolved, a, args, context)
+
```
Comment (dirk-thomas) [inline tools/roslaunch/src/roslaunch/substitution_args.py]: Unrelated change. Should be be reverted.

File: tools/roslaunch/src/roslaunch/xmlloader.py [hunk 9]
```diff
@@ -180,9 +181,15 @@ def resolve_args(self, args, context):
         Wrapper around substitution_args.resolve_args to set common parameters
         """
         # resolve_args gets called a lot, so we optimize by testing for dollar sign before resolving
+
```
Comment (dirk-thomas) [inline tools/roslaunch/src/roslaunch/xmlloader.py]: Unrelated change. Should be be reverted.

File: tools/roslaunch/src/roslaunch/xmlloader.py [hunk 10]
```diff
@@ -180,9 +181,15 @@ def resolve_args(self, args, context):
         Wrapper around substitution_args.resolve_args to set common parameters
         """
         # resolve_args gets called a lot, so we optimize by testing for dollar sign before resolving
+       
         if args and '$' in args:
+            #print("Args: %s"%(args))
```
Comment (dirk-thomas) [inline tools/roslaunch/src/roslaunch/xmlloader.py]: Unrelated change. Should be be reverted.

File: tools/roslaunch/src/roslaunch/xmlloader.py [hunk 11]
```diff
@@ -216,6 +223,7 @@ def reqd_attrs(self, tag, context, attrs):
 
     def _check_attrs(self, tag, context, ros_config, attrs):
         tag_attrs = tag.attributes.keys()
+
```
Comment (dirk-thomas) [inline tools/roslaunch/src/roslaunch/xmlloader.py]: Unrelated change. Should be be reverted.

File: tools/roslaunch/src/roslaunch/xmlloader.py [hunk 12]
```diff
@@ -289,10 +297,24 @@ def _arg_tag(self, tag, context, ros_config, verbose=True):
         Process an <arg> tag.
         """
         try:
+
```
Comment (dirk-thomas) [inline tools/roslaunch/src/roslaunch/xmlloader.py]: Unrelated change. Should be be reverted.

File: tools/roslaunch/src/roslaunch/xmlloader.py [hunk 13]
```diff
@@ -289,10 +297,24 @@ def _arg_tag(self, tag, context, ros_config, verbose=True):
         Process an <arg> tag.
         """
         try:
+            
             self._check_attrs(tag, context, ros_config, XmlLoader.ARG_ATTRS)
             (name,) = self.reqd_attrs(tag, context, ('name',))
-            value, default, doc = self.opt_attrs(tag, context, ('value', 'default', 'doc'))
+
```
Comment (dirk-thomas) [inline tools/roslaunch/src/roslaunch/xmlloader.py]: The extra empty line should be removed.

Comment (tahsinkose) [reply tools/roslaunch/src/roslaunch/xmlloader.py]: Removed.

File: tools/roslaunch/src/roslaunch/loader.py (hunk 4 above)
Comment (tahsinkose) [reply tools/roslaunch/src/roslaunch/loader.py]: Reverted.

File: tools/roslaunch/src/roslaunch/loader.py (hunk 3 above)
Comment (tahsinkose) [reply tools/roslaunch/src/roslaunch/loader.py]: Reverted.

File: ros_comm/package.xml (hunk 1 above)
Comment (tahsinkose) [reply ros_comm/package.xml]: Reverted.

File: tools/roslaunch/src/roslaunch/xmlloader.py (hunk 11 above)
Comment (tahsinkose) [reply tools/roslaunch/src/roslaunch/xmlloader.py]: Reverted.

File: tools/roslaunch/src/roslaunch/loader.py (hunk 3 above)
Comment (tahsinkose) [reply tools/roslaunch/src/roslaunch/loader.py]: Reverted.

File: tools/roslaunch/src/roslaunch/loader.py (hunk 2 above)
Comment (tahsinkose) [reply tools/roslaunch/src/roslaunch/loader.py]: Reverted.

File: tools/roslaunch/src/roslaunch/substitution_args.py (hunk 5 above)
Comment (tahsinkose) [reply tools/roslaunch/src/roslaunch/substitution_args.py]: Reverted.

File: tools/roslaunch/src/roslaunch/xmlloader.py (hunk 9 above)
Comment (tahsinkose) [reply tools/roslaunch/src/roslaunch/xmlloader.py]: Reverted.

File: tools/roslaunch/src/roslaunch/xmlloader.py (hunk 10 above)
Comment (tahsinkose) [reply tools/roslaunch/src/roslaunch/xmlloader.py]: Reverted.

File: tools/roslaunch/src/roslaunch/xmlloader.py (hunk 12 above)
Comment (tahsinkose) [reply tools/roslaunch/src/roslaunch/xmlloader.py]: Reverted.

**Comment by cwecht** [comment]
This PR requires  #1665 to be merged. The tests are failing because the tests are using mocks which are using lists instead of dictionaries.

Also the new substitution ``param`` is limited to arg-tags. This is but fine but one should be aware of this.

File: tools/roslaunch/src/roslaunch/xmlloader.py [hunk 14]
```diff
@@ -294,8 +299,19 @@ def _arg_tag(self, tag, context, ros_config, verbose=True):
         try:
             self._check_attrs(tag, context, ros_config, XmlLoader.ARG_ATTRS)
             (name,) = self.reqd_attrs(tag, context, ('name',))
-            value, default, doc = self.opt_attrs(tag, context, ('value', 'default', 'doc'))
             
+            params_restore_point = copy.deepcopy(context.params)
+            resolve_dict_restore_point = copy.deepcopy(context.resolve_dict)
+            # Use context to emulate get_param functionality in substitution_args
+            for key,v in ros_config.params.items():
+                p = Param(key,v)
```
Comment (cwecht) [inline tools/roslaunch/src/roslaunch/xmlloader.py]: ``ros_condfig.params`` is usually a map from ``str`` to  ``Param``, so you should be able to add `v` directly.

Comment (tahsinkose) [reply tools/roslaunch/src/roslaunch/xmlloader.py]: I'll look it in the weekend. Thanks for your comments.

**Comment by doisyg** [comment]
@tahsinkose Any update on this?
I have been using `param` as a substitution in `arg` tags using your fork for few months. It works great. I would love to see it merged and even better to see it work for the other tags

**Comment by dirk-thomas** [comment]
Waiting for #1665.

**Comment by dirk-thomas** [comment]
@tahsinkose Please update the target branch to `noetic-devel` which has been created since this ticket was created.

**Comment by tahsinkose** [comment]
> @tahsinkose Please update the target branch to `noetic-devel` which has been created since this ticket was created.

What about the back-compatibility for `melodic-devel`? I need this for `melodic-devel`.

**Comment by tahsinkose** [comment]
`deregister.test` of `test_rospy` seems to have failed in CI. However it passes locally. Can we have a rerun for debian stretch target @dirk-thomas?

**Comment by dirk-thomas** [comment]
> What about the back-compatibility for `melodic-devel`? I need this for `melodic-devel`.

Please see #1496 which describes the process. Only after a patch has been accepted to the latest branch it will be considered for backporting.

**Comment by tahsinkose** [comment]
> Only after a patch has been accepted to the latest branch it will be considered for backporting.

Should I close this then?
